docs/ember/authoring-addons.md

```diff
@@ -0,0 +1,53 @@
+Adding Glint support to TypeScript-based addons can provide a lot of value. 
+
+First, the addon itself will benefit from the additional typing coverage, as all templates in rendering tests or the dummy app will get type checked, thus increasing type safety and decreasing the likelihood of any type inconsistencies to sneak in over time (e.g. false positives in tests).
+
+Second, it makes consuming the addon for Glint-enabled apps much easier, providing proper Glint-compatible types out of the box.
+
+## Publishing addons with Glint types
+
+The first part is adding the proper Glint signature types to all the addon's components, helpers and modifiers, just the same way as described in these guides.
+
+If all apps were already using [first class component templates] (i.e. [strict mode] templates), this would be all it needs. But as this is certainly not the case for the time being, we will need to find a way to expose the required [Template Registry] entries to the consuming app, in an easy but also flexible way.
+
+As described in the [Using Addons] guide, the *easy* way for users is to import a file from the addon containing all the registry entries. This is for the most common case in which the app is consuming the addon provided components, helpers and modifiers as they are. The convention here is for the addon - when published on npm - to provide a `glint.d.ts` file, that the app will import as `import 'awesome-addon/glint'`. For a (v1) addon using [`ember-cli-typescript`][ect] this would be a `/addon/glint.ts` file, which needs to contain all the registry entries for all the public components (helpers, modifiers) the addon exposes to the app (in the addon's `/app` tree):
+
+{% code title="addon/glint.ts" %}
+
+```typescript
+import type ButtonComponent from 'awesome-addon/components/button';
+import type ModalComponent from 'awesome-addon/components/modal';
+
+declare module '@glint/environment-ember-loose/registry' {
+  export default interface Registry {
+    AwesomeButton: typeof ButtonComponent;
+    AwesomeModal: typeof ModalComponent;
```

Comment:
Uh, really? I was actually developing that specific habbit, to disambiguate between different things (components, models, services etc.) that might share the same base name. But also I am not willing to die on that hill! :wink:

Reply:
The reason it's *distinctly* not preferable for components is: when you *invoke* a component, you don't invoke it as `<AwesomeButtonComponent>`, you invoke it as `<AwesomeButton>`. Particular in a world with First-Class Component Templates and Glint-powered auto-import (…once we implement that 😉), you'll want to be able to type `<Awesome` and hit <kbd>^</kbd><kbd>␣</kbd> and have it autocomplete to `AwesomeButton` *not* to `AwesomeButtonComponent`.

Otherwise, imagine this:

```js
import AwesomeFormComponent from './awesome-form';
import AwesomeButtonComponent from './awesome-button';
import AwesomeInputComponent from './awesome-input';

<template>
  <AwesomeFormComponent as |form|>
    <AwesomeInputComponent @value={{@name}} @update={{@setName}} @label="Name" />
    <AwesomeButtonComponent @type='submit' @action={{form.submit}} />
  </AwesomeFormComponent>
</template>
```

😬 😬 😬 

This is clearly (to me at least!) very much preferable:

```js
import AwesomeForm from './awesome-form';
import AwesomeButton from './awesome-button';
import AwesomeInput from './awesome-input';

<template>
  <AwesomeForm as |form|>
    <AwesomeInput @value={{@name}} @update={{@setName}} @label="Name" />
    <AwesomeButton @type='submit' @action={{form.submit}} />
  </AwesomeForm>
</template>
```

Beyond components—it's quite reasonable to use that pattern to disambiguate models, services, controllers, etc., at least to a degree. (Routes and controllers are particularly in need of disambiguation even so, as you nearly always have matching pairs of them!) That said, I personally try to solve the overlap by giving things the best names that I can for the module they're coming from and then using modules as namespaces or *occasionally* by renaming it at the point of export. On the other hand, `Session` *might* be an object representing the session state separately from a `SessionService`, but unless it's actually ambiguous in my codebase, I would tend to name the service `Session` and give it a `state` property or some such. Then I would name the latter `State` inside that module and tak advantage of the fact that externally I can rename it on importing it if I *need* to disambiguate it:

```ts
import { type State as SessionState } from '../services/session';
```

Or, alternatively, I could do it when exporting it:

```ts
export { type State as SessionState };
```

But here I think it’s much more arguable than components, where the way we use them means it just basically *never* makes sense to have the suffix on them in my opinion!